Make Default Publisher Priority an updatable parameter (#1770)
moq-wg/moq-transport: draft-ietf-moq-transport.md

conditions that constitute a malformed track when detected by a receiver
include:

1.  An Object with a particular Subgroup ID is received, but its
     Publisher Priority is different from that of the previous Object with the same

>      Subgroup ID.

Reviewer: this is still true when the publisher priority wasn't updated.

Reviewer 2: This opens an interesting quesiton:

If I started a subgroup with no priority (inherit default) and the default changed mid-subgroup, do newly received objects in that subgroup get the new priority?  If the answer is _yes_, then deleting the above _is correct_.  

If I'm a relay, should I reprioritize the matching outgoing subgroups?  If so, we need to state that.

Author: > @suhasHere this is still true when the publisher priority wasn't updated.

It may be hard to tell if a publisher priority update happened, especially in multi-hop (even just P->R->S) or multi-publisher cases. 

Also it's now effectively a hop-by-hop compression parameter (maybe it always was?), so a Relay may decide to send an explicit per-object/per-subgroup priority instead of updating the default param. If we accept that priority changes are racy/best-effort, I think we have to drop any priority-related malformation rules, as the publisher now doesn't know what the subscriber (or anyone downstream) thinks the priority is for any given object that was sent without an explicit priority.

> @afrind If I'm a relay, should I reprioritize the matching outgoing subgroups? If so, we need to state that.

Good point, I hate this. 
How do you then expose this downstream for the next hops? 

I don't think we should actually allow effective mid-subgroup publisher priority changes in the current prioritization model. Priority should be fixed once you receive the start of a subgroup, and changes apply only to not-yet-started subgroups. This should be document in the document. 

I could see a mid-subgroup priority change being useful, but I think we would need a bigger update to the prioritization model. But reading the discussion here, we may need that anyway :/

`Publisher Priority` is a priority number associated with an individual
schedulable object.  A default for the subscription is specified in the
DEFAULT_PUBLISHER_PRIORITY parameter ({{publisher-priority}}). Publisher
priority can also be set per subgroup or datagram in the subgroup header or
datagram (see {{data-streams}}), which MUST take precedence over the default.

> DEFAULT_PUBLISHER_PRIORITY can be updated using REQUEST_UPDATE. Changing

Reviewer 2: Is this better in the DEFAULT_PUBLISHER_PRIORITY section, or here in priorities?

Reviewer 3: Priorities is already large, so I'd prefer to keep this in another section.

the default is best-effort and not synchronized with the data plane. When
an object omits the Publisher Priority, a subscriber resolves it using
the default in effect when the object was received and SHOULD cache
the resolved value, so a later change does not alter already-cached

> objects. If conflicting Publisher Priority values are observed for the

Reviewer 4: I'm a bit concerned about how this is detected and interaction with caches. Implementations tend to drop duplicate objects without much looking into them.  It might be better to say use which ever is seen first than lowest. I also don't want to imply there is any requirement for a relay to detect duplicates.

Author: Under "9.1. Caching Relays" we currently have 

> An endpoint that receives a duplicate Object with a different Forwarding Preference, Subgroup ID, Priority or Payload *MUST* treat the track as Malformed.

so this should not be something new. For non-caching relays, or generally in cases where we do not know that this is even conflicting value because we do not have the "old" object the "If conflicting Publisher Priority values are **observed**" doesn't apply anyway. 

The point of that statement was to have some recommended behavior, but I am happy to drop that.

> same object, a subscriber SHOULD use the lower priority.

Reviewer 2: is it clearer if we say "lower priority value"?

This paragraph contains important text, I'm just not sure where it should live.

Reviewer 2: Do we need to explain the possible race conditions and mitigations, or let people figure it out?

Reviewer 5: Expect more explanations about the race conditions. SHOULD is already a pretty strict requirement. Isn't there any other possibility here?

Author: **Placement.** I think this section is the right place to explain all the priorities handling. And the wire section is already quite heavy on intermingled logic.
 
**Races.** I did not try to implement synchronization based on https://github.com/moq-wg/moq-transport/issues/1270#issuecomment-4719569704
 
> Based on the discussions last week, I didn't hear folks asking for synchronization - sounds like the ones who want this feature were fine with some racy edge cases.
 
Also I did not find any easy way to spell this out without much more machinery required for a feature that seems niche.
 
> Do we need to explain the possible race conditions and *mitigations*
 
The REQUEST_UPDATE may arrive too early or too late.
 
I do not think there is currently a mitigation for "too early" without some kind of location in the update message, but that's more a rewrite of the whole REQUEST_UPDATE. Or a general look at priorities.
 
I think spelling out very explicitly "Changing the default is best-effort and not synchronized with the data plane" is the right level of calling out in normative text. It warns the reader that there is raciness, but enumerating all possible behaviors seems out-of-place for normative text describing this feature.
 
Of course this will be a different story if consensus emerges that we're unhappy with the raciness and want a robust solution.

Reviewer 3: We could allow the sender to put Largest Object in the REQUEST_UPDATE and say that if its present, the priority is effective as of Largest Object?

Reviewer 4: If applications care about the race, they could include priority in the object during the time of transition where the race can occur.

Author: > If applications care about the race, they could include priority in the object during the time of transition where the race can occur.

This covers the "too late" direction, but not REQUEST_UPDATE arriving *too early*.

When the publisher decides to change the priority, it can start stamping explicit priority on objects from that point on and send REQUEST_UPDATE. But objects sent before that decision, already in flight/submited to the QUIC stack can't be retroactively stamped. 

Since the bidi request stream is prioritized above data-plane, I’d expect this to be the most common form of the race.

> We could allow the sender to put Largest Object in the REQUEST_UPDATE and say that if its present, the priority is effective as of Largest Object?

Should this be a generic `REQUEST_UPDATE` synchronization primitive? Would we want to allow it only for this specific case?

Reviewer 3: It looks like SUBSCRIPTION_STATE_NOTIFY #1820 will land soon, in which case you could include Priority and the Largest Object/First Object with the new Priority there.  If we had that, I'm not sure we need to include it in the Objects as well?

If omitted from SUBSCRIBE, PUBLISH_OK or FETCH, the publisher uses
the value 128.

### DEFAULT PUBLISHER PRIORITY Parameter {#publisher-priority}

The DEFAULT_PUBLISHER_PRIORITY parameter (Parameter Type 0x0E)
is a uint8. It MAY appear in a PUBLISH, SUBSCRIBE_OK, FETCH_OK, or

> REQUEST_UPDATE message. It is the default Publisher Priority (see

Reviewer 2: ```suggestion
REQUEST_UPDATE message sent by the publisher on a subscription. It is the default Publisher Priority (see
```

Author: I am not sure what "on a subscription" brings/defines here?

Reviewer 2: It just scopes where this can occur.  The publisher can't send it in a REQUEST_UPDATE message to any other kind of request (eg PUB_NS)

{{priorities}}) for Subgroups and Datagrams of the subscription that
do not carry an explicit Publisher Priority. Lower numbers get higher

> priority.

Reviewer 2: I don't think we need to say how priorities work here, there's a ref to the section.

```suggestion
do not carry an explicit Publisher Priority.
```

SUBSCRIBE_NAMESPACE, SUBSCRIBE_TRACKS) can later send a REQUEST_UPDATE on the
same bidi stream as the request to modify it.  A subscriber can also send
REQUEST_UPDATE to modify parameters of a subscription established with PUBLISH.
The publisher can also send REQUEST_UPDATE to modify parameters set

> using SUBSCRIBE_OK.

Reviewer 3: ```suggestion
same bidi stream as the request to modify it.  A subscription can be modified by
either the Publisher or Subscriber using REQUEST_UPDATE, subject to the
constraints of the specific Parameter ({{message-params}}).
```

SUBSCRIBE_NAMESPACE, SUBSCRIBE_TRACKS) can later send a REQUEST_UPDATE on the
same bidi stream as the request to modify it.  A subscriber can also send
REQUEST_UPDATE to modify parameters of a subscription established with PUBLISH.

> The publisher can also send REQUEST_UPDATEto modify parameters set

Reviewer 2: ```suggestion
The publisher can also send REQUEST_UPDATE to modify parameters set
```

> using SUBSCRIBE_OK.

Reviewer 2: ```suggestion
using PUBLISH or SUBSCRIBE_OK.
```

Author: This is covered above by

> sender of a request (SUBSCRIBE, PUBLISH, FETCH, PUBLISH_NAMESPACE,
SUBSCRIBE_NAMESPACE, SUBSCRIBE_TRACKS) can later send a REQUEST_UPDATE

This whole paragraph doesn't read great, but I don't have a good suggestion how to improve it.

* Exclude your own tracks from SUBSCRIBE_NAMESPACE (#1596)
* Add Session-Level Tracks reserved namespace (#1562)
* Allow coalescing REQUEST_UPDATE processing (#1540)
* Make Default Publisher Priority an updatable parameter; stop malforming on

>   priority change (#1270)

Reviewer 2: We write the comprehensive release notes when we publish a new draft version.  The previous text is immutable.

```suggestion
```
